driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -215,23 +229,41 @@ private Throwable checkOutFailed(final Throwable t) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.TIMEOUT));
         } else if (t instanceof MongoOpenConnectionInternalException) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.CONNECTION_ERROR));
-            result = assertNotNull(t.getCause());
-        } else if (t instanceof IllegalStateException && t.getMessage().equals("The pool is closed")) {
+            result = t.getCause();
+        } else if (t instanceof MongoConnectionPoolClearedException) {
+            connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.CONNECTION_ERROR));
+        } else if (ConcurrentPool.isPoolClosedException(t)) {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.POOL_CLOSED));
         } else {
             connectionPoolListener.connectionCheckOutFailed(new ConnectionCheckOutFailedEvent(serverId, Reason.UNKNOWN));
         }
         return result;
     }
 
+    @Override
+    public void invalidate(@Nullable final Throwable cause) {
+        assertTrue(sdamProvider.optional().isPresent());
+        if (stateAndGeneration.pauseAndIncrementGeneration(cause)) {
+            LOGGER.debug("Invalidating the connection pool for " + serverId + " and marking it as 'paused'"
+                    + (cause == null ? "" : " due to " + cause.toString()));
+            openConcurrencyLimiter.signalClosedOrPaused();
+        }
+    }
+
     @Override
     public void invalidate() {
-        LOGGER.debug("Invalidating the connection pool");
-        generation.incrementAndGet();
-        connectionPoolListener.connectionPoolCleared(new ConnectionPoolClearedEvent(serverId));
+        invalidate(null);
+    }
+
+    @Override
+    public void ready() {
+        if (stateAndGeneration.ready()) {
+            LOGGER.debug("Marking the connection pool for " + serverId +  " as 'ready'");
+        }
     }
 
     public void invalidate(final ObjectId serviceId, final int generation) {
+        assertFalse(sdamProvider.optional().isPresent());
```

Comment:
TODO for @stIncMale : move `sdamProvider.optional().isPresent()` into a method with a descriptive name, e.g., `boolean loadBalanced()`.